### ISSUE TITLE
single node cluster deployment not permitted

### ISSUE THREAD

**Issue by ddiss** [SUSE/DeepSea#68]
In attempting to deploy a single node Ceph cluster, I ran into the following error during stage 1:
<pre>
658         priorities = []
659         for network in networks:
660             quantity = len(networks[network])
661             # Minimum number of nodes, ignore other networks
662             if quantity > 3:
663                 priorities.append( (len(networks[network]), network) )
664                 log.debug("Including network {}".format(network))
665             else:
666                 log.warn("Ignoring network {}".format(network))
667 
668         if not priorities:
669             raise ValueError("No network exists on at least 4 nodes")
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
</pre>

So it seems that in attempting to automagically determine which network is public and which network is cluster, Deapsea asserts that at least four nodes exist.

I only raise this as a bug because ceph-deploy supported single node cluster deployment. If this isn't seen as a priority to Deepsea then feel free to close this.

I'll gladly take this as bug assignee, but would appreciate an indication of how many other roadblocks I can expect to hit in attempting to support this.

**Comment by jan--f** [comment]
Of the top of my head I could imagine problems in several places:

- The validate runner at the sart of stage 3 is aimed at preventing you from doing this (among other things).
- The actual calls to start the daemons don't take any arguments as to what ports to use aso. so runing multiple daemons of the same kind won't work. This could probably be solved in a number of ways.
- The proposal runner doesn't handle it as you found out.

I wonder what @swiftgist thinks, but I wonder if it might be better to implement single node clusters in an extra set of state files. This could be less complex then the current stages, e.g. skip the hardware probing.

@ddiss do you envision a 'real' single node cluster with OSDs runing on disks or more something like a vstart cluster?

**Comment by ddiss** [comment]
@jan--f Thanks for the feedback.

I don't actually want to run more than one MDS or MON on the one node - just OSDs, so explicit port assignments shouldn't be needed. I'll take a look at stage 3 validation to see how intrusive the single-node-cluster changes there would be.

Regarding your last question: I doubt a single node cluster has any use outside of testing, but in my use case it'd be far superior to vstart because:
- I have a single node with seven disks, and want to make the best use of them for testing
- I'd prefer to have something closer to a customer setup without needing to up the node count

**Comment by swiftgist** [comment]
So, the original intent was to align with SES4 production requirements.  That is, all production clusters would have redundant monitors.  For the storage nodes, Ceph will still be happy when using replicas and the loss of one node.

Here are a few of the unsolved issues for running a single monitor:

* How do you prevent the addition of a single monitor (and end up with 2 monitors) in a user-friendly way?  This is in Stage 3.
* How do you prevent the removal of a single monitor (and end up with 2 monitors) in a user-friendly way?  This is in Stage 5.
* The check isn't in only the network detection.  It's also in the validate.py.

Is this a production vs. development setup?  If so, maybe a set of alternate defaults for developers would be the easiest (like Jan suggested).  Additionally, the above issues could be ignored since that would be an opt-in.

Stepping back, does DeepSea offer any advantages over vstart for the one monitor/one OSD cluster?  The overhead of configuring Salt (and possibly DeepSea) seems like overkill.

**Comment by smithfarm** [comment]
@jan--f Recently on the mailing list we discussed two use cases for DeepSea to be able to deploy single-node clusters:

1. unit testing of DeepSea itself
2. using DeepSea as a deployment tool for teuthology (quite a few teuthology tests run on one- or two-node clusters).

Making and maintaining a separate "mode" for this would seem to defeat the purpose and create more work?

What do we gain by preventing users from doing stuff? Do we think someone is going to complain because DeepSea allowed them to deploy a single-node cluster?

Could we make it so DeepSea emits a warning, but does what is asked? The advantages for testing are, I think, pretty obvious.

**Comment by smithfarm** [comment]
@swiftgist Prevent the addition of a second monitor? Why would you want to prevent that? If you're running a single-monitor cluster and want to upgrade to a 3+ monitor setup, you'd have to have 2 monitors at least for a short time...

**Comment by swiftgist** [comment]
@ddiss You answered my question as I sent it. :)
The shortest path may be the following since I did not have this in mind at the time.

* Add a method to populate.py that mimics proposals minus the undesired steps.  A good/nice name that this is intended for development would be ideal.
* Create a custom /srv/salt/ceph/stage/1/default.sls.  Maybe default-developer.sls.  Call populate.nice_name
* Create a custom /srv/salt/ceph/stage/3/default.sls.  Same idea (default-developers.sls).  Remove the validate call (starts on line 21).  Alternatively, do the same thing for validate.py that is done for populate.py

Take a look at https://github.com/SUSE/DeepSea/wiki/override#overriding-default-steps for adding variables to the pillar to use your custom files.

BTW, I wouldn't be too worried about trying to refactor too much.  I suspect this (variable validation) may be something larger that I didn't consider and those runners are in need of a reorganization.

**Comment by smithfarm** [comment]
This might not be obvious to everyone here, so I'll mention it - teuthology tests only run on "real" clusters. There is no option to use vstart.sh.

Currently, teuthology deploys its testing clusters via an Ansible recipe. We would like to get away from that and use DeepSea, but I guess it's only worth doing if it provides meaningful testing coverage to DeepSea.

**Comment by swiftgist** [comment]
@smithfarm I need to prevent the "I want a 2 monitor Ceph cluster as my final answer" issue.  I expect that it's unintentional, but if you only select 2 nodes as monitors, I need DeepSea to fail gracefully.  I do not have any additional checks other than the validation and the network discovery.

**Comment by smithfarm** [comment]
@swiftgist Why is that needed? 2 monitors is a valid Ceph configuration. If it's needed for oA, that sanity check could be done there...

**Comment by smithfarm** [comment]
My main point here is this: if we make a separate, special mode of DeepSea for deploying testing clusters, the tests will only cover that special mode (i.e. the ones customers are not using).

**Comment by swiftgist** [comment]
@smithfarm I thought 2 monitors could never establish quorum?  Honestly, I never tried it, but believed the propaganda.

If your main point is that we have to alter DeepSea to accommodate Teuthology, then Teuthology is using configurations that customers are not.  That may be fine for its purpose, but I do not think DeepSea will be able to accommodate customers, developers and testers with a single configuration.

**Comment by smithfarm** [comment]
DeepSea would not be accommodating teuthology - it would be the other way around. We would put effort into making teuthology use DeepSea because it would give DeepSea (and Salt) more testing coverage.

I also wanted to note that teuthology has tests that call for multiple MONs to run on a node, so if DeepSea doesn't allow that, that is another deal breaker.

I'm not a monitor expert, but the propaganda I read says any number of MONs is OK, although once you have 7 MONs, the incremental benefit of adding additional ones probably doesn't outweigh the additional overhead (law of diminishing returns). I have never deployed a 2-MON cluster, but I have deployed many 4-MON clusters and they worked fine.

**Comment by smithfarm** [comment]
Really, I don't understand why we are putting artificial limitations into DeepSea. If the idea is to keep users from doing stuff that is valid and allowed by Ceph, but isn't recommended/supported/"good for them", we should ask ourselves this question, first: "Will these clueless users be running DeepSea from the command line at all?"

Perhaps we could consider having a discussion on the topic of which validation checks makes sense to have in DeepSea and which ones are better placed in oA. 

Here's what would make sense to me: have DeepSea enable deployment of any cluster configuration that is allowed by Ceph itself, and put "policy" type checks into the front-end. I admit, though, that the name of the configuration file, `policy.cfg`, doesn't bode well for this idea.

**Comment by ddiss** [comment]
@swiftgist and @smithfarm thanks for the feedback. I'd be happy with a loud warning or developer parameter to enable deployment of a non-recommended setup.
Apparently @rjfd may have already implemented similar functionality for Openstack test automation, so I'll wait to hear from him before proceeding here. I'm fine using ceph-deploy for now.

**Comment by jan--f** [comment]
> Really, I don't understand why we are putting artificial limitations into DeepSea. If the idea is to keep users from doing stuff that is valid and allowed by Ceph, but isn't recommended/supported/"good for them", we should ask ourselves this question, first: "Will these clueless users be running DeepSea from the command line at all?"
Yes that was the intention. The motivation for this was due to the often less then helpful error messages from salt. Sometimes its ridiculously hard to figure out what went wrong, so the thought was to prevent misconfigurations.

**Comment by rjfd** [comment]
> Apparently @rjfd may have already implemented similar functionality for Openstack test automation, so I'll wait to hear from him before proceeding here. I'm fine using ceph-deploy for now.

@ddiss When I started the test automation prototype I tried to use only a single node to deploy DeepSea, but as you found out, DeepSea is not prepared to be deployed in a single node. So, I decided to use 4 nodes (the minimum required to pass all validation steps) in the deployment.

>So, the original intent was to align with SES4 production requirements.

My opinion on this matter is that DeepSea should not impose restrictions on the kind of configuration the user wants to deploy. I do understand the arguments given by @swiftgist regarding the requirements for the SES product, but what I think people haven't still fully realized is that DeepSea is an opensource project, which is trying to build a community, and should not be so strongly tied to SES.

In other words, DeepSea should not impose any restrictions to the Ceph configuration, such as the minimum number of MONs or OSDs, because of any requirements of any product where DeepSea is used.

In my view I see two solutions to this problem, either 
a) we have the upstream DeepSea code that does not impose any restrictions, and we have a downstream DeepSea fork where we diverge the code base to support product requirements; or 
b) we implement some kind of plugin system in the upstream code that will support the imposition of restrictions based on the OS distro that is used, with the ability to easily enabled/disable such plugins.

Both solutions have problems, and I'm not sure what would be the best.
Does anyone have another solution or wants to discuss the above solutions?

**Comment by theanalyst** [comment]
How about something like a deployment_type: dev sort of a switch in global which we can use to make validate happy for the time being?

**Comment by theanalyst** [comment]
similarly we should transition the 4 node check to a warning, everyone hand modifying the validate check doesn't seem like the right way out here, we could probably defer to the os based check for that in validate or  the plugin architecture @rjfd proposed.

**Comment by swiftgist** [comment]
Resolved with PR #141
